### ISSUE TITLE
High order zero coefficients are removed when creating Poly objects

### ISSUE THREAD

**Issue by iyanmv** [mhostetter/galois#197]
I found this inconsistency between NumPy and Galois while adapting my unit tests to `galois.Poly`, but maybe this behavior is on purpose by design so I wanted to ask here. It is somehow related to #196.

When you create a polynomial in NumPy, high order coefficients that are zero are always kept, unless the [`trim()`](https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.Polynomial.trim.html#numpy.polynomial.polynomial.Polynomial.trim) method is used. I found this behavior useful for error correction codes because the sizes of messages or codewords are fixed by the initial coefficients creating the polynomial, no matter if after the encoding/decoding operations (with polynomial operations) some high order coefficients are zero.

I write here a small example showing this difference

```python
>> import numpy as np
>> from numpy.polynomial import Polynomial as P
>> import galois
>> from galois import GF2

>> coeffs = [0, 1, 0, 0, 0, 0]
>> poly_numpy = P(coeffs)
>> poly_galois = galois.Poly(coeffs, field=GF2, order="asc")

>> poly_numpy.coef
array([0., 1., 0., 0., 0., 0.])

>> poly_numpy.trim().coef
array([0., 1.])

>> poly_galois.coeffs
GF([1, 0], order=2)
```

Is this the expected behavior? On the one hand, I get that `Poly` constructor uses the smallest possible array to store the coefficients. But on the other hand, if I want to replace NumPy polynomials objects with Galois `Poly` objects on encoders/decoders, now I always have to check and adapt (add missing zeros) the coefficients arrays that represent the messages/codewords of a code.

What do you think? Should `Poly` objects trim the coefficients during construction or add a method that does this when desired by the user like NumPy does?

**Comment by mhostetter** [comment]
It is done intentionally. Whether it *should* be done, is a separate question.

You are correct. It is done to only save / preserve the smallest array necessary. If a degree-400 polynomial were divided by a degree-300 polynomial, should the result have size 401? Often in this library, particularly when verifying primitivity or irreducibility of polynomials, taking polynomials to very higher powers and modding by a smaller polynomial are common. Without reducing intermediate results, I fear the *silent* memory consumption may be misleading, and wasteful.

There are some cases where you might like a fixed-size array of coefficients from a polynomial. I see two potential options:

1. Copy out the coefficients.
```python
x = np.zeros(32, dtype=int)  # Fixed-size array of length 32
# Create `poly`
# Do some arithmetic with `poly`
x[-len(poly):] = poly.coeffs  # Store the result in the pre-allocated array
```

2. Add a kwarg to return the coefficients in a fixed-size array.
```python
# Create `poly`
# Do some arithmetic with `poly`
x = poly.coeffs(size=32)  # Return the coefficients in an array of size 32, with zeros for higher-order terms
```

**Comment by mhostetter** [comment]
@iyanmv how does NumPy handle this case: if `x` is a degree-10 polynomial and `y` is a degree-5 polynomial, what is the size of the coefficients of `z = x**2 % y`?  `x**2` can have degree at most 20, but `x**2 % y` can have degree at most 5. Does it store 20 coefficients, or 5, or only those needed? Just curious.

**Comment by mhostetter** [comment]
I just did a little test. FWIW, it seems after performing arithmetic, NumPy does trim the result.

```python
In [1]: import numpy as np

In [2]: x = np.polynomial.polynomial.Polynomial([0, 1, 0, 0, 0])

In [3]: x
Out[3]: Polynomial([0., 1., 0., 0., 0.], domain=[-1,  1], window=[-1,  1])

In [4]: x.coef
Out[4]: array([0., 1., 0., 0., 0.])

In [6]: z = x**2; z
Out[6]: Polynomial([0., 0., 1.], domain=[-1.,  1.], window=[-1.,  1.])

In [7]: z.coef
Out[7]: array([0., 0., 1.])
```

**Comment by iyanmv** [comment]
> Does it store 20 coefficients, or 5, or only those needed?

Actually, in that case it just stores those needed. I guess NumPy is also not very consistent on this topic.

```python
>> from numpy.polynomial import Polynomial as P
>> x = P.basis(10)  # x^10
>> y = P.basis(5)  # x^5
>> z = x**2 % y
>> z.coef
array([0.])
```

Okay, so I agree that the library should always use the smallest possible array. Your memory argument convinced me.

**Comment by iyanmv** [comment]
> it seems after performing arithmetic, NumPy does trim the result.

Yes, exactly! I was also trying it out now.

**Comment by mhostetter** [comment]
I do think it's best to store the minimum coefficients necessary, but optionally returning more coefficients (if requested) is good / fine. I'm all for making users' lives easier.

I wish I could overload `Poly.coeffs` as a property and as a method `Poly.coeffs(size=None, order="desc")`, where `Poly.coeffs == Poly.coeffs()`. But someone could use a fixed size or different order by setting the kwargs in `Poly.coeffs()`.

**Comment by mhostetter** [comment]
@iyanmv do you think this issue is resolved given #210?

**Comment by iyanmv** [comment]
This is also fixed indeed!
